js.go

```diff
@@ -846,21 +846,228 @@ func (sub *Subscription) ConsumerInfo() (*ConsumerInfo, error) {
+type pullOptFn func(opts *pullOpts) error
+
+func (opt pullOptFn) configurePull(opts *pullOpts) error {
+	return opt(opts)
+}
+
+// PullNoWait enables receiving a no messages error from the server
+// in case there are no new messages when making a pull request.
+func PullNoWait() PullOpt {
+	return pullOptFn(func(opts *pullOpts) error {
+		opts.noWait = true
+		return nil
+	})
+}
+
+// PullExpiresIn defines when to let the server expire an inflight pull request.
+func PullExpiresIn(duration time.Duration) PullOpt {
+	return pullOptFn(func(opts *pullOpts) error {
+		opts.expires = duration
+		return nil
+	})
+}
+
+// PullMaxWaiting defines the max inflight pull requests to be delivered more messages.
+func PullMaxWaiting(n int) SubOpt {
+	return subOptFn(func(opts *subOpts) error {
+		opts.cfg.MaxWaiting = n
+		return nil
+	})
+}
+
+// Fetch pulls a batch of messages from a stream for a pull consumer.
+func (sub *Subscription) Fetch(batch int, opts ...PullOpt) ([]*Msg, error) {
+	if sub == nil {
+		return nil, ErrBadSubscription
+	}
+
+	var o pullOpts
+	for _, opt := range opts {
+		if err := opt.configurePull(&o); err != nil {
+			return nil, err
+		}
+	}
+
+	// Default is no wait, so async.
+	if o.ctx != nil && o.ttl != 0 {
+		return nil, ErrContextAndTimeout
+	}
+
 	sub.mu.Lock()
-	if sub.jsi == nil || sub.jsi.deliver != _EMPTY_ || sub.jsi.pull == 0 {
+	if sub.jsi == nil || sub.typ != PullSubscription {
 		sub.mu.Unlock()
-		return ErrTypeSubscription
+		return nil, ErrTypeSubscription
 	}
-	batch := sub.jsi.pull
-	nc, reply := sub.conn, sub.Subject
+
+	// NOTE: Subject only used to lookup to which stream the subject belongs to.
+	nc, _ := sub.conn, sub.Subject
 	stream, consumer := sub.jsi.stream, sub.jsi.consumer
 	js := sub.jsi.js
+
+	ttl := o.ttl
+	if ttl == 0 {
+		ttl = js.wait
+	}
 	sub.mu.Unlock()
 
-	req, _ := json.Marshal(&nextRequest{Batch: batch})
+	// Use the given context or setup a default one for the span
+	// of the pull batch request.
+	var (
+		ctx    = o.ctx
+		err    error
+		cancel context.CancelFunc
+	)
+	if o.ctx == nil {
+		ctx, cancel = context.WithTimeout(context.Background(), ttl)
+		defer cancel()
+	}
+
+	// Check if context not done already before making the request.
+	select {
+	case <-ctx.Done():
+		if ctx.Err() == context.Canceled {
+			err = ctx.Err()
+		} else {
+			err = ErrTimeout
+		}
+	default:
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// PullNoWait when enabled returns:
+	//
+	// - 404 right away if there are no more messages,
+	// - 409 in case it reached MaxAckPending.
+	//
+	// PullExpiresIn will give a duration to the server to expire our pull request
+	// to this inbox only and will serve the newest one.
+	//
+	// Default is to let the pull requests linger until there are 512 waiting requests,
+	// this limit can be modified by using PullMaxWaiting when creating the consumer.
+	//
+	nr := &nextRequest{Batch: batch, NoWait: o.noWait, Expires: o.expires}
+	req, _ := json.Marshal(nr)
```

Comment:
In case only Batch is set, could take advantage of the fast path and just the number as bytes.

Reply:
Also results in a lighter protocol.